### PR TITLE
Switch to untyped AST in the TinyLang.Field.Parser

### PR DESCRIPTION
**Pull request by jakzale**
We introduce a Raw (untyped) AST for `TinyLang.Field.Parser`

- fix some `haddock` errors, so that `tiny-lang` is searchable by `stack hoogle`,
- add `TinyLang.Field.Raw.Parser` for producing untyped AST (#47),
- refactor existential qualification tag to `TinyLang.Field.Existential`,
- rafactor `Uni`, `KnownUni`, `UniVal`, and `SomeUniVal` to `TinyLang.Field.UniVal`,
- generalise `Data.Field.TextField` typeclass to all parsers working on `String`,
- generalise `TinyLang.ParseUtils` to all parsers working on `String`,
- add `parseString` to `TinyLang.ParseUtils`.

### CODE REVIEW

File: field/TinyLang/Field/Generator.hs [hunk 1]
```diff
@@ -317,19 +317,17 @@ boundedArbitraryExprI
...
     => Vars f -> Int -> m (Expr f (AField f))
 boundedArbitraryExprI _    size | size <= 1 = EVal <$> arbitraryValI
+{-| NOTE.  If we allow variables here we won't generally know in
+  advance that they'll have integer values, so there would be a danger
+  that our comparisons will have a high probability of failing.  We
+  could fill the environment with lots of integer-valued variables to
+  reduce the risk of this, or supply a separate list of variables
+  which we're certain will only contain integer values.  This Note
+  also applies to the @size <= 1@ case above.
+-}
```
Comment (effectfully) [inline field/TinyLang/Field/Generator.hs]: That comment only applies to the `, (0, EVar <$> chooseUniVar vars)` line, so it shouldn't be moved away from it (but feel free to move it before the line rather than after it).

File: field/TinyLang/Field/UParser.hs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+type Parser = Parsec Void String
+
+type Identifier = String
+
+data Const
+  = CBool  Bool
+  | CInt   Integer
+  | CVec   [Bool]
+
+data Var = Var Identifier
```
Comment (effectfully) [inline field/TinyLang/Field/UParser.hs]: We don't need to define another `Var` here. Instead, `TinyLang.Var.Var` should be used.

File: field/TinyLang/Field/UParser.hs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                             , isDigit)
+
+type Parser = Parsec Void String
+
+type Identifier = String
+
+data Const
+  = CBool  Bool
+  | CInt   Integer
+  | CVec   [Bool]
```
Comment (effectfully) [inline field/TinyLang/Field/UParser.hs]: `TinyLang.Field.Core.SomeUniVal` should be used here. No point to introduce another definition.
Except raw terms shouldn't depend on typed terms, so `TinyLang.Field.Core.SomeUniVal` should be extracted somewhere common and both the typed and raw ASTs would depend on it.

File: field/TinyLang/Field/UParser.hs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+data Var = Var Identifier
+
+data Expr
```
Comment (effectfully) [inline field/TinyLang/Field/UParser.hs]: Should have the `f` parameter. `Integer` is not a thing at the moment (it may become later), currently we only have an abstract field.

File: field/TinyLang/Field/UParser.hs [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+lexeme = L.lexeme sc
+
+symbol :: String -> Parser String
+symbol = L.symbol sc
+
+
+-- Identifier Character
+isIdentifierChar :: Char -> Bool
+isIdentifierChar c =
+  or $ map (\f -> f c) [isLower , isDigit, (=='_'), (=='\'')]
```
Comment (effectfully) [inline field/TinyLang/Field/UParser.hs]: `or . map f` is better known as `any f`.

File: field/TinyLang/Field/UParser.hs [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+-}
+
+
+module TinyLang.Field.UParser where
+
+import           Control.Applicative.Combinators hiding (many, sepBy, some)
+import           Text.Megaparsec
+import           Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer      as L
+import           Data.Void
```
Comment (effectfully) [inline field/TinyLang/Field/UParser.hs]: We have `TinyLang.Prelude` that reexports a lot of stuff including `Data.Void`. It's much less annoying to develop things when you don't need to import `Data.Maybe`, `Data.List`, `Control.Monad`, etc all the time, so I'd suggest to replace `Data.Void` with `TinyLang.Prelude`, which also reexports `Set`, so you don't need to write the ugly looking `S.Set`.

File: field/TinyLang/Field/UParser.hs [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+import qualified Data.Set                        as S
+import           GHC.Unicode ( isLower
+                             , isDigit)
+
+type Parser = Parsec Void String
+
+type Identifier = String
+
+data Const
+  = CBool  Bool
```
Comment (effectfully) [inline field/TinyLang/Field/UParser.hs]: We use 4 spaces for indentation instead of 2 in Plutus, so I decided to keep that formatting rule in this project as well. Here's what I have in my `emacs` config:

```
(setq tab-stop-list (number-sequence 4 200 4))
(custom-set-variables
 '(haskell-indentation-layout-offset 4)
 '(haskell-indentation-left-offset 4)
 '(haskell-indentation-starter-offset 4)
 '(haskell-indentation-where-post-offset 4)
 '(haskell-indentation-where-pre-offset 4)
 '(tab-width 4)
 ...
)

File: field/TinyLang/Field/UParser.hs [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+
+-- Identifier Character
+isIdentifierChar :: Char -> Bool
+isIdentifierChar c =
+  or $ map (\f -> f c) [isLower , isDigit, (=='_'), (=='\'')]
+
+
+identifierChar :: (MonadParsec e s m, Token s ~ Char) => m (Token s)
+identifierChar = satisfy isIdentifierChar <?> "identifier character [a-z0-9_']"
+{-# INLINE identifierChar #-}
```
Comment (effectfully) [inline field/TinyLang/Field/UParser.hs]: Adding `INLINE` pragmas should be done together with benchmarking (I've seen cases where my code became slower after I sprinkled `INLINE` pragmas over it where I thought they would make sense), so don't bother adding this syntactic noise right now.

File: field/TinyLang/Field/UParser.hs [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+  The code is based on the tutorial at
+  https://markkarpov.com/megaparsec/parsing-simple-imperative-language.html
+
+  See also https://markkarpov.com/megaparsec/megaparsec.html
+-}
+
+
+module TinyLang.Field.UParser where
+
+import           Control.Applicative.Combinators hiding (many, sepBy, some)
```
Comment (effectfully) [inline field/TinyLang/Field/UParser.hs]: I thought `Text.Megaparsec` reexports the functions that are hidden here? If you remove `many` and `some`, will there be errors? What are they?

File: field/TinyLang/Field/UParser.hs [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+    ]
+
+isKeyword :: String -> Bool
+isKeyword = (`S.member` keywords)
+
+pIdentifier :: Parser String
+pIdentifier =
+  lexeme $ do
+  prefix     <- option "" (string "?" <|> string "#")
+  identifier <- (:) <$> lowerChar <*> many identifierChar
```
Comment (effectfully) [inline field/TinyLang/Field/UParser.hs]: `many` has terrible performance when it comes to parsing. Instead of `many (satisfy p)` (`identifierChar` expands to `satisfy`) use `takeWhileP`. [Proof](https://markkarpov.com/post/megaparsec-more-speed-more-power.html#there-is-hope).

We will eventually move from `String` to `Text`, so this is another reason to use `takeWhileP`.

File: field/TinyLang/Field/UParser.hs [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+pIntLiteral :: Parser String
+pIntLiteral =
+  lexeme (some digitChar)
+
+pVecLiteral :: Parser String
+pVecLiteral =
+  lexeme $
+  between
+  (symbol "{")
+  (symbol "}")
```
Comment (effectfully) [inline field/TinyLang/Field/UParser.hs]: Hard to read without any indentation.

File: field/TinyLang/Field/UParser.hs [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+  ((pBoolLiteral `sepBy` (symbol ",")) >>= (pure . concat))
```
Comment (effectfully) [inline field/TinyLang/Field/UParser.hs]: ``concat <$> (pBoolLiteral `sepBy` symbol ",")``?

Comment (jakzale) [reply field/TinyLang/Field/UParser.hs]: Yep, I am still getting used to applicative-style programming.

File: field/TinyLang/Field/Generator.hs (hunk 1 above)
Comment (jakzale) [reply field/TinyLang/Field/Generator.hs]: `haddock` version specified in `lts-13.26` breaks for that syntax.

Will check in spare time, if it was fixed in later version of `haddock`.

File: field/TinyLang/Field/UParser.hs (hunk 2 above)
Comment (jakzale) [reply field/TinyLang/Field/UParser.hs]: I was not sure about it.  `TinyLang.Var.Var` contains `Unique` field, which I believe is used to tell variables apart.

Since I am not checking for scopes at this stage, how should I generate values of `Unique`?

File: field/TinyLang/Field/UParser.hs (hunk 3 above)
Comment (jakzale) [reply field/TinyLang/Field/UParser.hs]: Do we want to have the untyped language and typed language depend on some common types?

File: field/TinyLang/Field/UParser.hs (hunk 8 above)
Comment (jakzale) [reply field/TinyLang/Field/UParser.hs]: :+1:

File: field/TinyLang/Field/UParser.hs (hunk 9 above)
Comment (jakzale) [reply field/TinyLang/Field/UParser.hs]: Ah, for some reason I thought that `between` was not in `Text.Megaparsec`.  Must have checked the wrong version of `megaparsec`.

File: field/TinyLang/Field/UParser.hs (hunk 3 above)
Comment (effectfully) [reply field/TinyLang/Field/UParser.hs]: Yes.

File: field/TinyLang/Field/Generator.hs (hunk 1 above)
Comment (effectfully) [reply field/TinyLang/Field/Generator.hs]: Don't bother, just remove `^` to make it a regular comment.

You found that out by running `haddock`? That's great.

File: field/TinyLang/Field/UParser.hs (hunk 2 above)
Comment (effectfully) [reply field/TinyLang/Field/UParser.hs]: In

```haskell
parseExpr
    :: forall f m. (MonadSupply m, TextField f)
    => String -> m (Either String (SomeUniExpr f))
parseExpr = fmap _scopedValue . parseScopedExpr
```

the `MonadSupply` constraint is responsible for generation of fresh `Unique`s.

In `common/TinyLang/ParseUtils.hs` we have these definitions:

```haskell
type Parser = ParsecT Void String (StateT Scope Supply)

-- | 'Scope' maps names onto 'Var's.
type Scope = Map String Var

data Scoped a = Scoped
    { _scopedScope :: Map String Var
    , _scopedValue :: a
    } deriving (Functor, Foldable, Traversable)

-- | Look up a variable name. If we've already seen it, return the corresponding Var;
-- otherwise, increase the Unique counter and use it to construct a new Var.
makeVar :: String -> Parser Var
makeVar name = do
    vars <- get
    case Map.lookup name vars of
        Just var -> pure var
        Nothing  -> do
            var <- freshVar name
            put $ Map.insert name var vars
            pure var
```

There are two options: either we keep these definitions or do not assign any `Unique`s to variables during parsing and instead do that in a separate pass, which forces us to parameterize the untyped AST by `var`, which gets instantiated to `String` during parsing and to `TinyLang.Var.Var` after the separate pass.

I'm slightly in favor of the latter option, because it also allows to instantiate `var` with a type of De Bruijn indices, which is handy for tests, compression, etc (we use that trick in Plutus). But the former option is fine too as it's fewer changes to make and transitioning from the former to the latter is a straightforward thing to do, so we can do that once we actually need it.

**Comment by effectfully** [comment]
Forgot to add: it's fine to keep the raw AST and the parser in the same module during development, but before merging the PR we'll need to separate the two ASTs and their infrastructure to different folders, so that we'll have the following file structure:

- `TinyLang.Field.Typed.Core`
- `TinyLang.Field.Typed.Evaluator`
- `TinyLang.Field.Raw.Core`
- `TinyLang.Field.Raw.Parser`
- etc

We'll drop the `Field` prefix once we get rid of the `boolean` language.

File: field/TinyLang/Field/UParser.hs [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+    where
+        isTF x = x == 'T' || x == 'F'
+        --
+        charToBool 'T' = True
+        charToBool 'F' = False
+        charToBool _   = error "impossible"
+
+pIntLiteral :: Parser Integer
+pIntLiteral =
+    lexeme $ fmap read (some digitChar)
```
Comment (effectfully) [inline field/TinyLang/Field/UParser.hs]: `some` is as inefficient as `many`.

File: field/TinyLang/Field/UParser.hs (hunk 4 above)
Comment (jakzale) [reply field/TinyLang/Field/UParser.hs]: Should the `Raw` AST be parameterised by `f`, or should the `f` be existentially quatified within the type?

File: field/TinyLang/Field/UParser.hs (hunk 3 above)
Comment (jakzale) [reply field/TinyLang/Field/UParser.hs]: Then I guess I should refactor out:
- `Uni`, and
- `SomeUniVal`.

As `SomeUniVal` depends on `Uni`.

File: field/TinyLang/Field/UParser.hs (hunk 4 above)
Comment (effectfully) [reply field/TinyLang/Field/UParser.hs]: Parameterized.

File: field/TinyLang/Field/UParser.hs (hunk 3 above)
Comment (effectfully) [reply field/TinyLang/Field/UParser.hs]: Yes.

File: field/TinyLang/Field/UParser.hs [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+      , binary (symbol  ">")      $ FGt
+      , unary  (keyword "neq0")   $ Neq0
+      ]
+    , [ binary (keyword "xor")    $ Xor
+      , binary (keyword "and")    $ And
+      , binary (keyword "or")     $ Or
+      ]
+    , [ unary  (keyword "unpack") $ Unp
+      ]
+    ]
```
Comment (jakzale) [inline field/TinyLang/Field/UParser.hs]: @effectfully, what is the operator precedence for this language?

The original spec mentions the following precedences:
- `not > xor > and > or`, and
- `{neg,inv} > {*,/} > {+,-}`

but it does not mention operator precedence for the following
- `neq0`
- `==, <=, <, >=, >`
- `unpack`
- `expr [ expr ]`

So far I assumed that:
- `not` > everything
- arithmetic > `neq0`,
- arithmetic > boolean, and 
- everything  > unpack.

Comment (effectfully) [reply field/TinyLang/Field/UParser.hs]: All prefix unary operators (`not`, `neq0`, `neg`, `inv`, `unpack`) should have the same precedence that is greater than the precedences of all the binary operators. And both `unOp1 (unOp2 x)` and `unOp1 (x binOp2 y)` always require parentheses. `unOp1 x binOp unOp2 y` never requires parentheses.

`expr [ expr ]` is a postfix operator. It should have precedence that is less than the one of unary operators, but still greater than the precedences of all the binary operators. So that `unOp x [ expr ]` never requires parentheses, but `(x binOp y) [ expr ]` always requires them.

Binary operators are handled as per usual like you've assumed.

Comment (jakzale) [reply field/TinyLang/Field/UParser.hs]: What is the difference between:
- `unOp1 (x binOp2 y)`, and
- `unOp1 x binOp2 y`
?

Comment (effectfully) [reply field/TinyLang/Field/UParser.hs]: Sorry for being unclear.

- `unOp1 (x binOp2 y)` parses as `unOp1 (x binOp2 y)`
- `unOp1 x binOp2 y` parses as `(unOp1 x) binOp2 y`

Comment (effectfully) [reply field/TinyLang/Field/UParser.hs]: Err,

> Implements the order specified by @effectfully, however it does permit
the following syntax `a + b [c]` where `[c]` is binding as a postfix operator to `b`.

by my "`(x binOp y) [ expr ]` always requires parentheses" I mean that if the user wants an expression to parse like that, then parentheses are required. If the user wants to parse an expression like `x binOp (y [ expr ])`, then parentheses are not required and both of

- `x binOp y [ expr ]`
- `x binOp (y [ expr ])`

are fine. This is what I wanted to say by

> expr [ expr ] is a postfix operator. It should have precedence that is <...> greater than the precedences of all the binary operators. 

And this same logic applies to other precedence-related rules that I mentioned.

So what you implement is exactly what is desired, there is no "however" there.

File: field/TinyLang/Field/Raw/Parser.hs [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+expressions nor the usual statements.
+
+@
+statement ::=
+    "let" var "=" expr
+    "assert" expr
+    "for" var "=" int "to" int "do" statements "end"
+
+statements ::=
+    (statement (";" statement)*)?
```
Comment (jakzale) [inline field/TinyLang/Field/Raw/Parser.hs:122]: @effectfully, I have a question about parsing `statements`:  should the grammar permit empty statements or empty expressions?

First of all, I think you already mentioned that the grammar is slightly akward, with statements being the other way round, and currently, the smallest program with a statement is
```
assert T; T
```
For for loops the original grammar from `Parser.hs` permits `statements` to be empty, we should be able to write
```
for x = 0 to 10 do end; T
```
but the AST does not permit the above.

Theoretically, we can permit empty loops.  Should I edit the grammar to permit them?

Also, neither grammar nor ast permit the following program:
```
assert T;
```

Comment (effectfully) [reply field/TinyLang/Field/Raw/Parser.hs:122]: > I have a question about parsing statements: should the grammar permit empty statements or empty expressions?

We don't want any empty expressions. We may want empty statements in the raw AST just for the ease of targeting and optimizing it (fewer invariants to preserve). We'll then need to remove them when mapping the raw AST to the typed one.

Permitting empty loops may be convenient for program generation (again, fewer invariants to worry about).

So yeah, I think should we allow empty loops. Please edit the grammar.

Oh, I totally forgot. Could you also add `for` loops to the AST and parse them directly instead of unrolling them on the fly like it was done previously? Like

```haskell
data Statement f
    = <...>
    | EFor Var (Expr f) (Expr f) [Statement f]
```

(a loop variable, a lower bound, an upper bound, a list of statements). Feel free to either do that here or in a separate PR depending on what is more convenient.

> Also, neither grammar nor ast permit the following program:
>
> `assert T;`

Yeah, that is something that we'll fix later.

File: field/TinyLang/Field/UParser.hs (hunk 2 above)
Comment (jakzale) [reply field/TinyLang/Field/UParser.hs]: Went with parameterising the AST by the type of the variable.

File: field/TinyLang/Field/Existential.hs [hunk 16]
```diff
@@ -0,0 +1,23 @@
+module TinyLang.Field.Existential
+    ( Some (..)
```
Comment (effectfully) [inline field/TinyLang/Field/Existential.hs:2]: Great that you've extracted this to a separate module. We'll eventually need to make a library out of the [much nicer stuff](https://github.com/input-output-hk/plutus/blob/master/language-plutus-core/src/Language/PlutusCore/Universe/Core.hs) that I've implemented in Plutus recently and you it here as well.

**Comment by jakzale** [comment]
@effectfully, I think this is getting pretty much done.

I will clean it up tomorrow, and then I will mark it ready for review.

File: common/TinyLang/ParseUtils.hs [hunk 17]
```diff
@@ -56,25 +58,35 @@ makeVar name = do
...
 
 -- 'parens' parses something between parenthesis.
-parens :: Parser a -> Parser a
+-- parens :: Parser a -> Parser a
+parens :: (MonadParsec e s m, Token s ~ Char, Tokens s ~ [Char]) => m a -> m a
 parens = between (symbol "(") (symbol ")")
 
-signedDecimal :: Integral a => Parser a
+brackets :: (MonadParsec e s m, Token s ~ Char, Tokens s ~ [Char]) => m a -> m a
+-- brackets :: Parser a -> Parser a
```
Comment (effectfully) [inline common/TinyLang/ParseUtils.hs]: Drop the comments, they outdate quickly.

File: field/TinyLang/Field/Raw/Core.hs [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+    = ELet    v          (Expr v f)
+    | EAssert (Expr v f)
+    | EFor    v          (Expr v f) (Expr v f) [Statement v f]
+
+{-| = Utility Type Aliases
+-}
+type RawExpr = Expr Var
+type RawStatement = Statement Var
+
+{-| = Type Instances
```
Comment (effectfully) [inline field/TinyLang/Field/Raw/Core.hs]: `type instance` is what you write for a clause of an open type family. I think it's common to call the things below just "instances" (or "class instances" or "type class instances").

File: field/TinyLang/Field/Raw/Parser.hs [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+analysis and grammar a bit.  To avoid any ambiguities between lexical
+analysis and grammar we will explicitly mark all lexical tokens.
+
+Please note that expressions are the designated start production for
+the parser.
+
+We are using the following convention
+
+* /tokens/ are enclosed in parentheses @""@.
+* /regular expressions/ are enclosed in square brackets @[]@.
```
Comment (effectfully) [inline field/TinyLang/Field/Raw/Parser.hs]: I think e.g. `[0-9]+` is a regular expression as a whole and `[0-9]` is just a part of it. I'd probably just drop this line.

File: field/TinyLang/Field/Raw/Parser.hs [hunk 20]
```diff
@@ -0,0 +1,339 @@
+{-|
+= Raw Parser
+
+A simple parser for language involving booleans, field elements, and
+vectors.
+
+Given that the language is simple enough we are mixieng lexical
```
Comment (effectfully) [inline field/TinyLang/Field/Raw/Parser.hs]: Typo: "mixieng".

File: field/TinyLang/Field/Raw/Parser.hs [hunk 21]
```diff
@@ -0,0 +1,339 @@
...
+statements ::=
+    (statement (";" statement)*)?
+@
+
+== Operator Precedence
+Precedence: "not" > "xor" > "and" > "or" (but use parentheses anyway).
+if-then-else has to be parenthesised unless it's at the very top.
+
+Precedence for numeric operators is standard:  {neg,inv} > {*,/} > {+,- }.
+Things like "neg inv 5" are illegal: use parentheses.
```
Comment (effectfully) [inline field/TinyLang/Field/Raw/Parser.hs]: The operator precedence needs to be either fixed (to include all the possible case that we talked about before: `unOp x binOp y`, etc) or just dropped.

File: field/TinyLang/Field/Raw/Parser.hs [hunk 22]
```diff
@@ -0,0 +1,339 @@
...
+    ]
+
+pExpr :: TextField f => Parser (RawExpr f)
+pExpr = Comb.makeExprParser pTerm operatorTable
+
+pTop :: TextField f => Parser (RawExpr f)
+pTop = top pExpr
+
+{-| == Helper functions
+-}
```
Comment (effectfully) [inline field/TinyLang/Field/Raw/Parser.hs]: I don't think we should export them from here. `parseRational` is too specific and unsafe, if the caller needs that, they can implement it themselves. Let's instead export a function that receives a parser and a string and outputs an `Either String` (or `Either Text` depending on what is more convenient). Without hardcoding the field to `Rational`.

File: field/TinyLang/Field/Raw/Parser.hs [hunk 23]
```diff
@@ -0,0 +1,339 @@
...
+
+parseRational :: String -> String -> String
+parseRational fileName str =
+    either
+      errorBundlePretty
+      show
+      $ runParser (pTop @Rational) fileName str
+
+p :: String -> IO ()
+p s = putStrLn $ parseRational "" s
```
Comment (effectfully) [inline field/TinyLang/Field/Raw/Parser.hs]: Let's not ever export one-letter names (or even create them on the top level, unless it's a quasi-quote or some other very special case).

File: library/Data/Field.hs [hunk 24]
```diff
@@ -98,14 +98,15 @@ makeDiv d x y = do
 fromDivided :: Maybe f -> f
 fromDivided = fromMaybe $ throw DivideByZero
 
-parseFieldDefault :: Field f => Parser f
+-- parseFieldDefault :: Field f => Parser f
```
Comment (effectfully) [inline library/Data/Field.hs]: Drop the comment.

File: test/Field/Raw/golden/operator-precedence-00.field [hunk 25]
```diff
@@ -0,0 +1,12 @@
...
+	- c * d
+    ==
+	  e / f
+	+ g / h
+xor
+	  a * b
+	- c * d
+    >=
+	  e / f
+	+ g / h
```
Comment (effectfully) [inline test/Field/Raw/golden/operator-precedence-00.field:11]: It would be really nice to have another file alongside these two where we pretty-print the result of parsing to see how things get formatted. But that can be a separate PR.

File: test/Field/Raw/Textual.hs [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+import System.FilePath
+import System.FilePath.Glob
+
+testDir :: FilePath
+testDir = "test" </> "Field" </> "Raw" </> "golden"
+
+gen_test_parsing :: IO TestTree
+gen_test_parsing =
+    do
+        files <- testFiles
```
Comment (effectfully) [inline test/Field/Raw/Textual.hs:28]: So annoying that you can't do `IO` in `TestTree`.

File: test/Main.hs [hunk 27]
```diff
@@ -15,4 +16,12 @@ test_all =
         ]
 
 main :: IO ()
-main = defaultMain test_all
+main =
+    do  test_rawParsing <- Raw.gen_test_parsing
```
Comment (effectfully) [inline test/Main.hs]: Why not alter `test_all` above (and change its type to `IO TestTree`) instead of inlining everything here?

File: field/TinyLang/Field/Raw/Core.hs [hunk 28]
```diff
@@ -0,0 +1,71 @@
...
+    | Div
+    | BAt
+
+data UnOp
+    = Not
+    | Neq0
+    | Neg
+    | Inv
+    | Unp
+
```
Comment (effectfully) [inline field/TinyLang/Field/Raw/Core.hs:57]: I'd write `deriving (Show)` here instead of having a standalone deriving declaration at the bottom as the latter results in a lot of boilerplate when there are many instances to derive.

Applies to other data types as well.

**Review by effectfully — APPROVED** [review]
LGTM. Comments are mostly nitpicking.

File: field/TinyLang/Field/Raw/Parser.hs [hunk 29]
```diff
@@ -0,0 +1,339 @@
...
+
+import           TinyLang.Prelude hiding ( option
+                                         , many
+                                         , try
+                                         )
+
+import           TinyLang.Field.Raw.Core
+import           TinyLang.Field.Existential
+import           TinyLang.Field.UniVal
+import           Data.Set ( fromList
```
Comment (effectfully) [inline field/TinyLang/Field/Raw/Parser.hs]: We have a convention that imports from external libraries and imports of internal modules are separated by a newline. The latter go first and the former go next (after a newline). This makes it easier to read imports.

File: test/Field/Raw/golden/operator-precedence-00.field (hunk 25 above)
Comment (jakzale) [reply test/Field/Raw/golden/operator-precedence-00.field:11]: I will add a pretty-printing golden testing pass as part of #30.

File: test/Field/Raw/Textual.hs (hunk 26 above)
Comment (jakzale) [reply test/Field/Raw/Textual.hs:28]: Yes.  I am afraid that was intentional:
https://stackoverflow.com/a/33046723

File: field/TinyLang/Field/Raw/Parser.hs (hunk 22 above)
Comment (jakzale) [reply field/TinyLang/Field/Raw/Parser.hs]: Ok,
moving `parseRational` to `test/Field/Raw/Textual`.

Added `TinyLang.ParseUtils.parseString` which evaluates to `Either String a`.

**Comment by jakzale** [comment]
@effectfully, incorporated your feedback.

Do you prefer to merge or squash-merge?

**Comment by effectfully** [comment]
> Do you prefer to merge or squash-merge?

The latter unless some particular care was taken to organize commits to be well-structured, self-contained, etc.
